heron/scheduler-core/src/java/com/twitter/heron/scheduler/RuntimeManagerMain.java

```diff
@@ -347,14 +347,18 @@ protected boolean validateRuntimeManage(
       return false;
     }
 
-    // Check whether cluster/role/environ matched
-    ExecutionEnvironment.ExecutionState executionState = adaptor.getExecutionState(topologyName);
-    if (executionState == null
-        || !executionState.getCluster().equals(Context.cluster(config))
-        || !executionState.getRole().equals(Context.role(config))
-        || !executionState.getEnviron().equals(Context.environ(config))) {
-      LOG.severe("cluster/role/environ not matched");
-      return false;
+    // Check whether cluster/role/environ matched. It is possible to get into a state where some of
+    // the data does not actually exist in zookeeper. If we are in this state them assume the
+    // runtime is valid.
+    if (adaptor.doesExecutionStateExist(topologyName)) {
```

Comment:
I see what you're saying now. So the idea would be that if for some reason the execution state was gone then some commands (restart, active, deactivate) would just simply fail and would require the topology to be killed and then submitted it order to use those commands?

 A larger change that could be made would be to heron itself write the execution state. The topology master could write the execution state as an ephemeral node in zookeeper so that its always there if the topology is running. This would fix the edge case around deleting that piece of state from zookeeper and might simplify things a bit.